### PR TITLE
Work around problems deserializing graphs on wasm32-unknown-unknown

### LINKED ISSUE #12: usize does not work well for de/serialization.
WASM is 32bit always (afaik).

So if you serialize a FastGraph on a 64bit platform the resulting bytes will not be deserializable in a WASM binary.

Is there a real need for `usize` or would you be amenable to a more concrete type? Or do you know of some way to tell serde that `usize` should read 8 bytes (`u64`) when it's decoding into a `u32`?

### PR DESCRIPTION
**Pull request by dabreegster**
Fixes #12. This is a breaking API change; old serialized graphs won't load with this new code, so this warrants a semantic version bump.

### CODE REVIEW

**Comment by easbar** [comment]
Thanks! Is there some easy way this can be made configurable? This way one could choose between using this (to achieve wasm compatibility) and using full 64bit for large graphs. Ideally we would also range check the data that is written if some values exceed `INVALID_NODE` etc.

**Comment by dabreegster** [comment]
Woops, I forgot the doc updates -- added.

> Is there some easy way this can be made configurable?

I can't think of any easy way. One idea is to check if the target environment is wasm at build-time and use different values for the constants, but maybe people are serializing CHs in a native env and reading them later in wasm. (This is my use case.) I think there'd have to be options passed to InputGraph to pick the max ID, then lots of plumbing to replace the constants.

> Ideally we would also range check the data that is written if some values exceed INVALID_NODE etc.

`add_edge` in `InputGraph` looks like the perfect place for this check. What should the response be -- panicking, returning Option<usize> for the edge ID, something else?

**Comment by Carterj3** [comment]
I think you could use a feature-flag to control `u32::max_value()`, `usize::max_value()`, or etc.
https://doc.rust-lang.org/1.29.0/book/first-edition/conditional-compilation.html

While I personally think enums are better for recording "Invalid Edges" than a magic number I think given that you are using magic numbers it's fair for an end user to have to select a maximum for their situation. Configurable magic numbers is also a much smaller change to implement.

My use case is the same as yours where I use a build script to produce some bytes that WASM will just deserialize (saves several minutes of time). I think most people using WASM would want to do it this way.

**Comment by easbar** [comment]
> I think you could use a feature-flag to control 

Ok, but does this also work when using fast paths as a dependency or only when compiling it standalone? I have not tried it but to me it sounds like the latter is true. Is it common to publish different flavors of a crate (like fast_paths_32 and fast_paths_64)?

> While I personally think enums are better for recording "Invalid Edges" than a magic number

fast paths uses integer arrays to store the graph and there has to be some notion of 'this entry does not exist', so one of the integer values in the range of the chosen integer type has to be distinguished. How can this be done with enums / why would this be better?

Right now I think the minimal change needed for wasm would be writing a little de/serialization method that 'knows' which special value is used for 'invalid edge' etc. (e.g. converting usize::max to u32::max and serializing everything into a wasm compatible 32 bit representation). The maybe better approach might be using 0 instead of max for the special value (this way the integer size would no longer matter).

**Comment by Carterj3** [comment]
So I've made a few examples. 

# [Features](https://github.com/easbar/fast_paths/compare/master...Carterj3:feature)
Features are set by your crate afaik. Although I don't think anything stops you from submitting multiple crates. I haven't seen that but I also don't spend a lot of time looking at crates.

The idea being that all of your crates would specify the same feature so they all thinks the maximums are `u32::max_value()`. (or if you don't care, default to `usize::max_value()`.

## Dependencies
When specifying fast_paths as a dependency you can just accept the default `fast_paths = "0.1.1`.

For this specific implementation you'd need to disable defaults in order for your crate to compile the dependency as say `u32` but doing `fast_paths = { version = "0.1.1", default-features = "false", features = ["invalid_u32"] }`

## Command Line
If you want to set the feature for some command line tests you can do so as follows.

`cargo test --release --no-default-features --features "invalid_usize" -- --nocapture run_performance_test_dist`
> number of nodes (input graph) ..... 40461
number of edges (input graph) ..... 85090
preparation time .................. 559 ms
number of nodes (fast graph) ...... 40461
number of out-edges (fast graph) .. 66520
number of in-edges  (fast graph) .. 66589
total query time .................. 1541 ms
query time on average ............. 15 micros
test tests::run_performance_test_dist ... ok

`cargo test --release --no-default-features --features "invalid_u64" -- --nocapture run_performance_test_dist`
> number of nodes (input graph) ..... 40461
number of edges (input graph) ..... 85090
preparation time .................. 549 ms
number of nodes (fast graph) ...... 40461
number of out-edges (fast graph) .. 66520
number of in-edges  (fast graph) .. 66589
total query time .................. 1480 ms
query time on average ............. 14 micros
test tests::run_performance_test_dist ... ok

`cargo test --release --no-default-features --features "invalid_u32" -- --nocapture run_performance_test_dist`
> number of nodes (input graph) ..... 40461
number of edges (input graph) ..... 85090
preparation time .................. 560 ms
number of nodes (fast graph) ...... 40461
number of out-edges (fast graph) .. 66520
number of in-edges  (fast graph) .. 66589
total query time .................. 1537 ms
query time on average ............. 15 micros
test tests::run_performance_test_dist ... ok

# [Enums](https://github.com/easbar/fast_paths/compare/master...Carterj3:Enum)
I like enums over magic numbers just because it feels more technically correct although in this specific case with magic numbers you'd never run into an issue unless you had exactly `usize` nodes/edges so it's not a realistic issue.

I tried to leave as much as EdgeId/NodeId as I could since if a piece of data is always going to be valid it's counterproductive to keep passing around the enum.

## Codesize
Only the [`replaced_{in,out}_edge`](https://github.com/Carterj3/fast_paths/blob/Enum/src/fast_graph.rs#L95) are enums so the impact is less than I expected.
So for each `FastGraphEdge` you have its size should be 7*usize instead of 5*usize.

## Performance
It has pretty similar performance to the non-enum versions (above) although this is all fairly unscientific.

> number of nodes (input graph) ..... 40461
number of edges (input graph) ..... 85090
preparation time .................. 578 ms
number of nodes (fast graph) ...... 40461
number of out-edges (fast graph) .. 66520
number of in-edges  (fast graph) .. 66589
total query time .................. 1600 ms
query time on average ............. 16 micros
test tests::run_performance_test_dist ... ok

**Comment by easbar** [comment]
> Right now I think the minimal change needed for wasm would be writing a little de/serialization method that 'knows' which special value is used for 'invalid edge' etc. (e.g. converting usize::max to u32::max and serializing everything into a wasm compatible 32 bit representation).

Why would this not be feasible? Using different crate variations also seems like possible approach, but I am not sure if it warrants the extra complexity.

@Carterj3 Thanks for the other changes you suggested, if I find the time I might go through them and refresh my rust knowledge a bit. But this seems not directly related to the wasm issue we are facing here?

**Comment by dabreegster** [comment]
@Carterj3 Thanks for trying out some more alternatives! I think the best choice is the enums branch. It's great that the performance impact is minimal! Any time somebody updates code involving IDs, the type system forces them to acknowledge the possibility that a node ID hasn't been set. Do you think Option<NodeId> would be a bit more clear than a new enum?

If the enums branch is too invasive a change, then the argument why using a special sentinel value as invalid is OK here: the value is contained within the scope of this crate; external users should never get INVALID_NODE back through the API.

The Features branch would be my second choice; it's strictly an improvement on this PR, by allowing people to opt into wasm compatibility or not. If we go with that option, there are probably a few comments we can add in the README to explain what to choose.

> Right now I think the minimal change needed for wasm would be writing a little de/serialization...

This sounds kind of complex to me. If you want people to have the option of using 2^64 nodes, you still need to plumb through the ability to make a choice, either by changing the API to accept options, or by using feature flags. If you don't care about providing that option and want a limit of 2^32 nodes, then this PR does that in a much simpler way.

tldr: My vote is for a variation of the enums branch (using Option<NodeId> and adding comments to the README). The features branch is my second vote.

**Comment by easbar** [comment]
Honestly, I think I do not get it. Let me try to review this issue. Fast Paths uses integer arrays to store, well, integers. They might range from 0 to 2^32-1 or 2^64-1 depending on whether the system in use resolves `usize` to a 32bit or a 64bit integer. The larger the graph is the more of these numbers need to be distinguished. For most graphs 32bit will be enough, only for very large ones 64bit will be needed. In some cases one has to pre-allocate these integer arrays and there are some entries in these arrays that either have no meaning yet or will never have a meaning. Since every entry in these arrays has to have a certain value I chose the `std::usize::MAX` to represent these 'special/gap' entries. Its a very simple concept and it could probably be done the exact same way in basically every programming language. 

Now the only issue with this approach so far seems that if one uses FastPaths on one machine where `usize` resolves to 64bit, serializes the graph to disk, and then opens the same graph on another device  where `usize` resolves to 32bit we would be stuffing 64bit integer values into 32bit integer variables and that obviously does not work. If the graph is actually so large that some of the array entries exceed 2^32-1 this approach inherently fails and there is no way to open a 64bit graph on a 32bit device. Otherwise it can be done and the only thing that we need to take care of is converting the maximum integer values (2^64-1) to 2^32-1. Why is this complex? To me it seems this is cannot be done any simpler. 

If we access one of the integer arrays, yes we could return Option to make it easier to handle the missing/special values. But how is this even related to the wasm issue?

> If you want people to have the option of using 2^64 nodes, 

Yes, I think I would prefer to keep the 64bit option, because the wasm use-case does not seem to justify restricting the entire crate to 32bit altogether.

> you still need to plumb through the ability to make a choice, either by changing the API to accept options, or by using feature flags.

Sorry I do not understand this. I am talking about adding a simple serialization method that only works as long as no integer (except the sentinel) exceeds 2^32-1. People who plan to use the serialized graph on a 32bit device can use it, others can use the standard 64bit version of the same code. Or maybe the problem here is that the serde crate works by annotating the data classes and there is no explicit function that does the serialization?

**Comment by Carterj3** [comment]
> Sorry I do not understand this. I am talking about adding a simple serialization method that only works as long as no integer (except the sentinel) exceeds 2^32-1. People who plan to use the serialized graph on a 32bit device can use it, others can use the standard 64bit version of the same code. Or maybe the problem here is that the serde crate works by annotating the data classes and there is no explicit function that does the serialization?

I don't think you can simply write some `fn parse(bytes: &[u8]) -> usize` that when it's given an `std::u64::MAX` returns a `std::u32::MAX` and tell Serde to use it.
I am no expert with Serde but from looking at a de/serializer [bincode](https://github.com/servo/bincode/blob/master/src/de/mod.rs), a manually implemented De/Serialize struct [linked-hash-map](https://github.com/contain-rs/linked-hash-map/blob/master/src/serde.rs), and the docs for [Serde](https://serde.rs/impl-deserialize.html) I'm not too sure where even to start. Perhaps you see a way to make a MVP for this but I don't.


> Now the only issue with this approach so far seems that if one uses FastPaths on one machine where usize resolves to 64bit, serializes the graph to disk, and then opens the same graph on another device where usize resolves to 32bit we would be stuffing 64bit integer values into 32bit integer variables and that obviously does not work.

So, I've learned a few things since initially writing the issue but in short the only reason `fast_paths` is an issue on WASM for me is because an `INVALID_EDGE` which may be stored in a `FastGraphEdge. replaced_in_edge` or a `FastGraphEdge. replaced_out_edge` has a value larger than `std::u32::MAX`. The rest of the `EdgeId`s and `NodeId`s in a `FastGraph` are always valid.

`bincode` can actually de/serialize `u64`<->`u32` so long as it's a valid value (ex. less than `std::u32::MAX`. (See this [playground](https://gist.github.com/rust-play/8e92411ad863a9faffb8e67d76facbd0) which doesn't run because you can't use bincode online I guess, but you can run it locally). So somehow the fact that `bincode` serialized a value to 8 bytes on a 64bit system isn't enough to cause a problem when using `bincode` to deserialize that value to only 4 bytes (32bit WASM).

---
> But this seems not directly related to the wasm issue we are facing here?

I'll try to restate a bit more explictly.

The [enum branch](https://github.com/easbar/fast_paths/compare/master...Carterj3:Enum) solves this issue for WASM because it doesn't store any `INVALID_EDGE`s in a `FastGraph` so there is no number larger than 2^32 in my graphs.
I personally like the Enum solution more than the Feature one but the Feature branch is certainly a less invasive change and could very easily be reverted if needed in the future.

The [feature branch](https://github.com/easbar/fast_paths/compare/master...Carterj3:feature) solves this issue for WASM because it allows me to choose between `u32`, `u64`, and `usize` for what to use as a maximum value. So, when I create a `InputGraph` and convert+serialize it as a `FastGraph` I use the feature for `invalid_u32` and my graph has no values larger than 2^32 which means WASM can deserialize it.
If somebody was going to make say a desktop client they could leave the `INVALID_{NODE,EDGE}` as `usize` and de/serialize without any 2^32 limitation.

**Comment by easbar** [comment]
> I don't think you can simply write some fn parse(bytes: &[u8]) -> usize that when it's given an std::u64::MAX returns a std::u32::MAX and tell Serde to use it.
I am no expert with Serde but from looking at a de/serializer bincode, a manually implemented De/Serialize struct linked-hash-map, and the docs for Serde I'm not too sure where even to start. Perhaps you see a way to make a MVP for this but I don't.

Ok I do not know either, but all we need to do is write a `Vec<usize>` to disk and thereby replace every `std::usize::MAX` with `std::u32::MAX` and throw an error if any entry is `>= std::u32::MAX`. That really cannot be that difficult. If there is really no other way we could even do exactly that: Convert the fast path graph to another datastructure that uses `u32` instead of `usize` and then serialize this instead of the actual graph (with the downside of requiring extra memory to create the converted graph).

> So, I've learned a few things since initially writing the issue but in short the only reason fast_paths is an issue on WASM for me is because an INVALID_EDGE which may be stored in a FastGraphEdge. replaced_in_edge or a FastGraphEdge. replaced_out_edge has a value larger than std::u32::MAX. The rest of the EdgeIds and NodeIds in a FastGraph are always valid.

I think its a little more than that: Every usize in `FastGraph` could potentially exceed `std::u32::MAX` so to properly store it and open it on a 32bit device later all these usize numbers have to be converted in the way I described above.

> The enum branch solves this issue for WASM because it doesn't store any INVALID_EDGEs in a FastGraph so there is no number larger than 2^32 in my graphs.

Like I said above its not only about `INVALID_EDGE`. `FastGraph#ranks` or `FastGraphEdge#weight` are also `usize` and might not be deserializable to `u32`. But lets ignore this for a moment, then my question is stil if the enum branch does not store `std::usize::MAX`, what does it store *then*? Are we talking about using/serializing `Vec<Enum<usize>>` or `Vec<Option<usize>` instead of `Vec<usize>`? And maybe somehow serde manages to serialize/deserialize `Option=None` between 64bit and 32bit systems? If its storing, say, 4byte numbers it has to reserve one value for the `None` case (just like we are using the `max` value so far). To me its seems very arcane to use `Vec<Option<usize>>` just because this way we take advantage of some special serialization handling of `Option=None` implemented by serde (which is even only needed to maintain compatibility between 32 and 64 bit systems) to be honest. But if this turns out to be the best/only viable option to support WASM now I would at least understand *why* this even *is* a solution.

> The feature branch solves this issue for WASM because it allows me to choose between u32, u64, and usize for what to use as a maximum value. So, when I create a InputGraph and convert+serialize it as a FastGraph I use the feature for invalid_u32 and my graph has no values larger than 2^32 which means WASM can deserialize it.
If somebody was going to make say a desktop client they could leave the INVALID_{NODE,EDGE} as usize and de/serialize without any 2^32 limitation.

I get the idea, but isn't the problem that the integer arrays in `FastGraph` are storing indices into other integer arrays and Rust always uses `usize` for these?  I think this was the reason I used `usize` instead of `u32` in the first place. If that was not a problem it could even be fine to always use `u32`, because the graphs in use are probably very rarely so large that they exceed the four byte limit (and in all other cases this way we could save 50% memory for the graphs). Just to give an example: As of now four bytes is still easily enough to store the world-wide OpenStreetMap road network (and maybe by the time this becomes a problem there is 64bit support for wasm as well).

tldr; I still think it would be easiest to add a 32bit serialization method.

**Comment by easbar** [comment]
I added the helper methods I had in mind here: https://github.com/easbar/fast_paths/pull/14.
Does this fix the issue for you @dabreegster, @Carterj3?

**Comment by dabreegster** [comment]
Sorry for the delayed response.

Re: the enum branch. I haven't done the legwork of trying out this, so I'm not sure all of the places where an invalid node/edge might get stored. But just to clarify, the reason this is more "technically correct" is about type-safety. In Rust (and also in languages like Java 8, now that there are Option types), representing some kind of invalid case as a sentinel value like -1 or MAX is a bit messy, because any code touching that value has to know about the special case. If it's an Option<usize>, then the compiler forces everyone to acknowledge the edge case.

I left some comments on #14 -- I think it's a good option too, but I don't see it as any more/less complicated than the other alternatives here yet.

**Comment by easbar** [comment]
Thanks a lot for taking a look and giving your opinion! 

> I'm not sure all of the places where an invalid node/edge might get stored. 

I think I spotted a few places where there are still usize (for example ranks and weights) that also would have to be turned into enums. 

> But just to clarify, the reason this is more "technically correct" is about type-safety. In Rust (and also in languages like Java 8, now that there are Option types), representing some kind of invalid case as a sentinel value like -1 or MAX is a bit messy, 

Well, in Java you would certainly not use a `List<Optional<Integer>>` or even some kind of `List<YourEnumHere>` if you are writing similar code as fast paths. Actually you would not even want to use `List<Integer>`, for performance reasons. But this is because of the way its implemented in Java and I assume in Rust this is less of a problem. Using an enum or Option we can use the compiler to enforce checking the availability of nodes, edges etc. ok, this might be an advantage in some cases, but to me this also seems a bit like over-complicating things. Having an array of numbers and knowing that in some (not even all) cases I need to check their value does not appear to be such a complex concept that really justifies using an enum/option to me. Furthermore, this whole issue we are discussing here is not even about type-safety. To me it still looks like the enum branch uses enums just because as a side effect we get the 64/32bit compatibility we are looking for (because of the way serde treats the `None` value)?

> I left some comments on #14 -- I think it's a good option too, but I don't see it as any more/less complicated than the other alternatives here yet.

Cool, thanks! If #14 fixes the issue for you this would still be my preferred way to fix the wasm issue.

**Comment by easbar** [comment]
Don't want to push, but any more thoughts on this @Carterj3?

**Comment by dabreegster** [comment]
Abandoning this in favor of the FastGraph32 approach just merged
